src/context.jl

```diff
@@ -1,24 +1,77 @@
 # OpenCL.Context
 
+const _ctx_reference_count = Dict{CL_context, Int}()
+
+
+function create_jl_reference!(ctx_id::CL_context)
+    if haskey(_ctx_reference_count, ctx_id) # for the first jl reference, we already have a refcount of 1
+        @check api.clRetainContext(ctx_id) # increase internal refcount, if creating an additional reference
+    end
+    refcount = get!(_ctx_reference_count, ctx_id, 0)
+    _ctx_reference_count[ctx_id] = refcount + 1
+    return
+end
+function free_jl_reference!(ctx_id::CL_context)
+    if !haskey(_ctx_reference_count, ctx_id)
+        error("Freeing unknown context")
+    end
+    refcount = _ctx_reference_count[ctx_id]
+    if refcount == 0
+        error("Double free of context id: ", ctx_id)
+    elseif refcount == 1
+        delete!(_ctx_reference_count, ctx_id)
+        return
+    end
+    _ctx_reference_count[ctx_id] = refcount - 1
+    return
+end
+
 type Context <: CLObject
```

Comment:
Well, but this is reference counting for correctness and not performance, so it should be fine!
It's just tracking the creation of the finalizer, so acting whenever the finalizer get's called is the right time!
Singleton doesn't sound too bad! We just need to figure out, how this works with different devices and if there are use cases, when people want multiple contexts on the same device...